mobile_verifier/src/speedtests_average.rs

```diff
@@ -0,0 +1,478 @@
+        {
+            id = report.pubkey.as_ref().to_vec(); // eww!
+            sum_upload += report.upload_speed;
+            sum_download += report.download_speed;
+            sum_latency += report.latency;
+            window_size += 1;
+        }
+
+        if window_size > 0 {
+            let upload_speed_avg_bps = sum_upload / window_size as u64;
+            let download_speed_avg_bps = sum_download / window_size as u64;
+            let latency_avg_ms = sum_latency / window_size as u32;
+            let validity = validity(
+                window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+            );
+            let tier = SpeedtestTier::new(
+                window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+            );
+            let reward_multiplier = tier.into_multiplier();
+            SpeedtestAverage {
+                pubkey: id.into(),
+                window_size: window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+                validity,
+                reward_multiplier,
+            }
+        } else {
+            SpeedtestAverage {
+                pubkey: id.into(),
+                window_size: 0,
+                upload_speed_avg_bps: sum_upload,
+                download_speed_avg_bps: sum_download,
+                latency_avg_ms: sum_latency,
+                validity: proto::SpeedtestAvgValidity::TooFewSamples,
+                reward_multiplier: Decimal::ZERO,
+            }
+        }
+    }
+}
+
+impl SpeedtestAverage {
+    pub async fn write(
+        &self,
+        filesink: &FileSinkClient,
+        speedtests: Vec<Speedtest>,
+    ) -> file_store::Result {
+        filesink
+            .write(
+                proto::SpeedtestAvg {
+                    pub_key: self.pubkey.clone().into(),
+                    upload_speed_avg_bps: self.upload_speed_avg_bps,
+                    download_speed_avg_bps: self.download_speed_avg_bps,
+                    latency_avg_ms: self.latency_avg_ms,
+                    timestamp: Utc::now().encode_timestamp(),
+                    speedtests: speedtests_without_lapsed(
+                        speedtests.iter(),
+                        Duration::hours(SPEEDTEST_LAPSE),
+                    )
+                    .map(|st| proto::Speedtest {
+                        timestamp: st.report.timestamp.timestamp() as u64,
+                        upload_speed_bps: st.report.upload_speed,
+                        download_speed_bps: st.report.download_speed,
+                        latency_ms: st.report.latency,
+                    })
+                    .collect(),
+                    validity: self.validity as i32,
+                    reward_multiplier: self.reward_multiplier.try_into().unwrap(),
```

Comment:
guaranteed safe unwrap?

Reply:
I believe so and copy pasted from original, so always been there.